src/main/java/com/example/mybookshopapp2/controllers/BooksRestApiController.java

```diff
@@ -0,0 +1,122 @@
+import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.methodOn;
+
+@RestController
+@RequestMapping("/api/books")
+//@Api(description = "book data api")
+public class BooksRestApiController {
+
+    private final BookService bookService;
+
+    @Autowired
+    public BooksRestApiController(BookService bookService) {
+        this.bookService = bookService;
+    }
+
+    @GetMapping(value = "/{bookId}")
+    public RepresentationModel<Book> getBookById(@PathVariable Integer bookId) {
+        return bookService.getBookById(bookId)
+                .add(
+                        linkTo(
+                                methodOn(BooksRestApiController.class)
+                                        .getBestSellerBooks())
+                                .withSelfRel());
+    }
+
+    @GetMapping(value = "/by-author-first-name", produces = "application/hal+json")
+    @ApiOperation("operation to get book list of bookshop by author")
+    public CollectionModel<Book> getBooksByAuthorFirstName(@RequestParam("author") String authorName) {
+
+        List<Book> booksByAuthor = bookService.getBooksByAuthorFirstName(authorName);
+        booksByAuthor.forEach(book -> {
+            book.add(
+                    linkTo(
+                            methodOn(BooksRestApiController.class)
+                                    .getBookById(book.getId()))
+                            .withSelfRel());
+        });
+        return CollectionModel.of(booksByAuthor);
+    }
+
+    @GetMapping("/by-title")
+    @ApiOperation("get books by book title")
+    @ApiResponses(value = {
+            @io.swagger.annotations.ApiResponse(code = 200, message = "Successful request"),
+
+    })
+    public RepresentationModel<?> booksByTitle(@RequestParam("title") String title) throws BookstoreApiWrongParameterException {
+        ApiResponse<Book> response = new ApiResponse<>();
+        List<Book> data = bookService.getBooksByTitle(title);
+        data.forEach(book -> {
+            book.add(
+                    linkTo(
+                            methodOn(BooksRestApiController.class)
+                                    .getBookById(book.getId()))
+                            .withSelfRel());
+        });
+        response.setDebugMessage("successful request");
+        response.setMessage("data size: " + data.size() + " elements");
+        response.setStatus(HttpStatus.OK);
+        response.setTimeStamp(LocalDateTime.now());
+        response.setData(data);
+        return RepresentationModel.of(ResponseEntity.ok(response));
+    }
+
+    @GetMapping("/by-price-range")
+    @ApiOperation("get books by price range from min price to max price")
+    public ResponseEntity<List<Book>> priceRangeBooks(@RequestParam("min") Integer min, @RequestParam("max") Integer max) {
+        return ResponseEntity.ok(bookService.getBooksWithPriceBetween(min, max));
+    }
+
+    @GetMapping("/with-max-discount")
```

Comment:
Please stick to the single code style and naming. Previous methods you called with by prefix. by-price.. , by-title etc. But this you called with-max-discount. Isn't it's better to call it by-max-discount?

Reply:
This code was written by the professor of the video course.